### PR TITLE
trinsic-cli: init at 1.1.0

### PR DESCRIPTION
**Pull request by tmarkovski**
###### Motivation for this change

Command line interface for Trinsic Ecosystems

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/admin/trinsic-cli/default.nix [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  unpackPhase = ''
+    tar xvzf $src
+    ls -la trinsic-cli
+  '';
+
+  buildPhase = ''
+    cd trinsic-cli
+    cargo vendor
+    cargo build --release
+  '';
```
Comment (legendofmiracles) [inline pkgs/tools/admin/trinsic-cli/default.nix:26]: Can be replaced by changing the directory in preBuild. And letting the defaults compile the program - so no need for running cargo manually

File: pkgs/tools/admin/trinsic-cli/default.nix [hunk 2]
```diff
@@ -0,0 +1,42 @@
+{ lib, stdenv, rustPlatform, fetchurl }:
+
+stdenv.mkDerivation rec {
+  pname = "trinsic-cli";
+  version = "1.1.0";
+
+  src = fetchurl {
+    url = "https://github.com/trinsic-id/sdk/releases/download/v${version}/${pname}-1.1.2.tar.gz";
```
Comment (legendofmiracles) [inline pkgs/tools/admin/trinsic-cli/default.nix:8]: Is there anything the release tarball offers over the raw source? Also, do you happen to know what the -nix tarball is?

File: pkgs/tools/admin/trinsic-cli/default.nix [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  '';
+
+  meta = with lib; {
+    description = "Trinsic CLI";
+    longDescription = ''
+      Command line interface for Trinsic Ecosystems
+    '';
+    homepage = "https://trinsic.id/";
+    license = licenses.asl20;
+    maintainers = with maintainers; [ ];
```
Comment (legendofmiracles) [inline pkgs/tools/admin/trinsic-cli/default.nix:40]: Maintainers needs to be set

File: pkgs/tools/admin/trinsic-cli/default.nix [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  };
+
+  nativeBuildInputs = [
+    rustPlatform.rust.cargo
+    rustPlatform.rust.rustc
+  ];
+
+  unpackPhase = ''
+    tar xvzf $src
+    ls -la trinsic-cli
```
Comment (legendofmiracles) [inline pkgs/tools/admin/trinsic-cli/default.nix:19]:
```suggestion
```

**Review by legendofmiracles — CHANGES_REQUESTED** [review]
Accidentally typed something here, and now I can't remove this comment, please ignore 😅

File: pkgs/tools/admin/trinsic-cli/default.nix (hunk 2 above)
Comment (tmarkovski) [reply pkgs/tools/admin/trinsic-cli/default.nix:8]: Custom `build.rs` file removed, which depends on `protoc` compiler - only used as dev tool. The languages for other bindings are excluded in the `-nix` tarball and also a missing Cargo.lock was included. I uploaded some manual packages that fix these issues.

File: pkgs/tools/admin/trinsic-cli/default.nix [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+  meta = with lib; {
+    description = "Trinsic CLI";
+    longDescription = ''
+      Command line interface for Trinsic Ecosystems
+    '';
+    homepage = "https://trinsic.id/";
+    license = licenses.asl20;
+    maintainers = with lib.maintainers; [ tmarkovski ];
+    platforms = [ "x86_64-linux" "x86_64-darwin" ];
```
Comment (legendofmiracles) [inline pkgs/tools/admin/trinsic-cli/default.nix:26]: Any reason why this wouldnot run on non x86 architectures?
```suggestion
```
It's generally set by buildRustPackage.

File: pkgs/tools/admin/trinsic-cli/default.nix [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  src = fetchFromGitHub {
+    rev = "v${version}";
+    owner = "trinsic-id";
+    repo = "sdk";
+    sha256 = "sha256:04lq2yv8ifh75wl63g03yh7rw6i83vsrn5g6injqjlsa4g4c928d";
+  };
+
+  cargoSha256 = "b932fe848c9400f803cc242c6febd3119f3e2b5e4bdb1ed0ef9100b811e217fb";
+
+  preBuild = "cd ./cli";
```
Comment (legendofmiracles) [inline pkgs/tools/admin/trinsic-cli/default.nix:16]: If we are no longer extracting the source manually then it's cleaner to use `sourceRoot = "cli"`

**Comment by legendofmiracles** [comment]
Result of `nixpkgs-review pr 146458` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>trinsic-cli</li>
  </ul>
</details>

**Comment by legendofmiracles** [comment]
All package related commits should be squashed into the first, and adding the maintainer to the general maintainers file should be called `maintainers: add tmarkovski`.

Unfortunately when running the program I'm getting weird errors:
```
$ ./results/trinsic-cli/bin/trinsic wallet
thread 'main' panicked at 'Unrecognized command', src/parser/wallet.rs:35:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by tmarkovski** [comment]
> All package related commits should be squashed into the first, and adding the maintainer to the general maintainers file should be called `maintainers: add tmarkovski`.
> 
> Unfortunately when running the program I'm getting weird errors:
> 
> ```
> $ ./results/trinsic-cli/bin/trinsic wallet
> thread 'main' panicked at 'Unrecognized command', src/parser/wallet.rs:35:9
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> ```
> 
> The config and help subcommands work.

Thanks @legendofmiracles, I'll address the noted. The correct wallet subcommand is `trinsic wallet create --name legendofmiracles`, you can then follow it with `trinsic wallet search`. Other command `trinsic config --show`.

File: pkgs/tools/admin/trinsic-cli/default.nix [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+  src = fetchurl {
+    url = "https://github.com/trinsic-id/sdk/releases/download/v${version}/trinsic-cli-vendor-${version}.tar.gz";
+    sha256 = "7e5377e8460ebb1253b9974e02d0077ffc58f2b4712cf7896f1bef7e9f580cd4";
+  };
+
+  cargoVendorDir = "vendor";
+  doCheck = false;
+
+  buildInputs = lib.optionals stdenv.isDarwin [ darwin.Security ];
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/trinsic-cli/default.nix:15]:
```suggestion
  buildInputs = lib.optionals stdenv.isDarwin [ Security ];
```

File: pkgs/tools/admin/trinsic-cli/default.nix [hunk 8]
```diff
@@ -0,0 +1,26 @@
+{ lib, stdenv, rustPlatform, fetchurl, darwin }:
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/trinsic-cli/default.nix:1]:
```suggestion
{ lib, stdenv, rustPlatform, fetchurl, Security }:
```

File: pkgs/top-level/all-packages.nix [hunk 9]
```diff
@@ -22691,6 +22691,8 @@ with pkgs;
 
   trinity = callPackage ../os-specific/linux/trinity { };
 
+  trinsic-cli = callPackage ../tools/admin/trinsic-cli { };
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:22694]:
```suggestion
  trinsic-cli = callPackage ../tools/admin/trinsic-cli {
    inherit (darwin.apple_sdk.frameworks) Securit
  };
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Review by legendofmiracles — APPROVED** [review]

**Comment by tmarkovski** [comment]
@SuperSandro2000 I'd appreciate your second look on this PR. Requested changes are in. Thanks

**Pull request merged**